gtsfm/retriever/netvlad_retriever.py

```diff
@@ -0,0 +1,262 @@
+"""Retriever implementation which provides a NetVLAD global image descriptor to suggest potential image pairs.
+
+Note: Similarity computation based off of Paul-Edouard Sarlin's HLOC:
+Reference: https://github.com/cvg/Hierarchical-Localization/blob/master/hloc/pairs_from_retrieval.py
+https://openaccess.thecvf.com/content_cvpr_2016/papers/Arandjelovic_NetVLAD_CNN_Architecture_CVPR_2016_paper.pdf
+
+Authors: John Lambert
+"""
+
+import math
+import os
+from dataclasses import dataclass
+from pathlib import Path
+from typing import List, Optional, Tuple
+
+import dask
+import matplotlib.pyplot as plt
+import numpy as np
+import torch
+from dask.delayed import Delayed
+
+import gtsfm.utils.logger as logger_utils
+from gtsfm.loader.loader_base import LoaderBase
+from gtsfm.retriever.retriever_base import RetrieverBase
+from gtsfm.frontend.cacher.global_descriptor_cacher import GlobalDescriptorCacher
+from gtsfm.frontend.global_descriptor.netvlad_global_descriptor import NetVLADGlobalDescriptor
+
+
+logger = logger_utils.get_logger()
+
+PLOT_SAVE_DIR = Path(__file__).parent.parent.parent / "plots"
+
+MAX_NUM_IMAGES = 10000
+
+
+@dataclass
+class SubBlockSimilarityResult:
+    i_start: int
+    i_end: int
+    j_start: int
+    j_end: int
+    subblock: torch.Tensor
+
+
+class NetVLADRetriever(RetrieverBase):
+    def __init__(self, num_matched: int, blocksize: int = 50) -> None:
```

Comment:
is this class specific to netvlad? 
Can this be done with other descriptors too? maybe ones @travisdriver was looking into?

Reply:
if you think it can be made generic, please create another issue before merging.

Reply:
This class is specific to NetVLAD